core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -521,402 +185,92 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
-    if (mCurrentBlockInStream != null && mCurrentBlockInStream.getId() == blockId
-        && mCurrentBlockInStream.remaining() != 0) {
-      return;
-    }
-    checkCacheStreamInSync();
-    updateBlockInStream(blockId);
-    if (PASSIVE_CACHE_ENABLED) {
-      updateCacheStream(blockId);
-    }
-  }
-
-  /**
-   * Updates {@link #mCurrentCacheStream}. When {@code mShouldCache} is true, {@code FileInStream}
-   * will create an {@code BlockOutStream} to cache the data read only if the block is read from a
-   * remote worker and we have an available local worker. Note that when the block is in UFS but not
-   * in Alluxio, the cache stream is not needed in the client because the worker caches directly the
-   * block when it reads from UFS. And if the block is already in local worker, then the block
-   * should not be cached again.
-   *
-   * The following preconditions are checked inside:
-   * <ol>
-   * <li>{@link #mCurrentCacheStream} is either done or null.</li>
-   * <li>EOF is reached or {@link #mCurrentBlockInStream} must be valid.</li>
-   * </ol>
-   * After this call, {@link #mCurrentCacheStream} is either null or freshly created.
-   * {@link #mCurrentCacheStream} is created only if the block is not cached in a chosen machine and
-   * position is at the beginning of a block. This function is only called by
-   * {@link #updateStreams()}.
-   *
-   * @param blockId the block ID
-   */
-  private void updateCacheStream(long blockId) {
-    // We should really only close a cache stream here. This check is to verify this.
-    Preconditions.checkState(mCurrentCacheStream == null || mCurrentCacheStream.remaining() == 0);
-    closeOrCancelCacheStream();
-    Preconditions.checkState(mCurrentCacheStream == null);
-
-    if (blockId < 0) {
-      // End of file.
-      return;
-    }
-    Preconditions.checkNotNull(mCurrentBlockInStream, "mCurrentBlockInStream");
-
-    // do not create cache stream when the block is not in remote worker
-    if (!(mShouldCache && mCurrentBlockInStream.getSource() == BlockInStreamSource.REMOTE)) {
+    if (mBlockInStream == null) { // no current stream open, advance position
+      mPosition = pos;
       return;
     }
 
-    // Unlike updateBlockInStream below, we never start a block cache stream if mPos is in the
-    // middle of a block.
-    if (mPositionState.getPos() % mBlockSize != 0) {
-      return;
-    }
-
-    try {
-      // If this block is read from a remote worker, we should never cache except to a local worker.
-      WorkerNetAddress localWorker = mContext.getLocalWorker();
-      if (localWorker != null) {
-        mCurrentCacheStream =
-            mBlockStore.getOutStream(blockId, getBlockSize(mPositionState.getPos()), localWorker,
-                mOutStreamOptions);
-      }
-    } catch (IOException e) {
-      handleCacheStreamException(e);
+    long delta = pos - mPosition;
+    if (delta <= mBlockInStream.getPos() && delta >= -mBlockInStream.remaining()) { // within block
```

Comment:
shouldn't this be `delta <= mBlockInStream.remaining() && delta >= -mBlockInStream.getPos()`?

Reply:
Good catch, I made this change after disabling the `FileInStream` tests which should catch this.